src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1978,183 +1987,205 @@ BuildJobTreeTaskList(Job *jobTree)
+			subqueryTask->jobId = jobId;
+			sqlTaskList = lappend(sqlTaskList, subqueryTask);
+		}
 
-			RangeTableFragment *shardFragment = palloc0(fragmentSize);
-			shardFragment->fragmentReference = shardInterval;
-			shardFragment->fragmentType = CITUS_RTE_RELATION;
-			shardFragment->rangeTableId = tableId;
+		++taskIdIndex;
+	}
 
-			tableSize += ShardLength(shardInterval->shardId);
+	return sqlTaskList;
+}
 
-			if (tableId == 1)
-			{
-				List *fragmentCombination = list_make1(shardFragment);
-				fragmentCombinationList = lappend(fragmentCombinationList,
-												  fragmentCombination);
-			}
-			else
-			{
-				List *fragmentCombination = (List *) lfirst(fragmentCombinationCell);
-				fragmentCombination = lappend(fragmentCombination, shardFragment);
 
-				/* get next fragment for the first relation list */
-				fragmentCombinationCell = lnext(fragmentCombinationCell);
-			}
-		}
+/*
+ * SubqueryTaskCreate creates a modify task by replacing the target
+ * shardInterval's boundary value when allRelationsJoinedOnPartitionKey
+ * is true. Then performs the normal shard pruning on the subquery.
+ *
+ * The function errors out if the subquery is not router select query (i.e.,
+ * subqueries with non equi-joins.).
+ */
+static Task *
+SubqueryTaskCreate(Query *originalQuery, ShardInterval *shardInterval,
+				   bool allRelationsJoinedOnPartitionKey,
+				   RelationRestrictionContext *restrictionContext,
+				   uint32 taskIdIndex)
+{
+	Query *copiedQuery = copyObject(originalQuery);
+
+	uint64 shardId = shardInterval->shardId;
+	Oid distributedTableId = shardInterval->relationId;
+	StringInfo queryString = makeStringInfo();
+	ListCell *restrictionCell = NULL;
+	Task *subqueryTask = NULL;
+	List *selectPlacementList = NIL;
+	uint64 selectAnchorShardId = INVALID_SHARD_ID;
+	List *relationShardList = NIL;
+	uint64 jobId = INVALID_JOB_ID;
+	bool routerPlannable = false;
+	bool upsertQuery = false;
+	bool replacePrunedQueryWithDummy = false;
+	bool allReferenceTables = restrictionContext->allReferenceTables;
+	RelationRestrictionContext *copiedRestrictionContext =
+		CopyRelationRestrictionContext(restrictionContext);
+	List *shardOpExpressions = NIL;
+	RestrictInfo *shardRestrictionList = NULL;
 
-		/*
-		 * Determine anchor table using shards which survive pruning instead of calling
-		 * AnchorRangeTableId
-		 */
-		if (anchorRangeTableId == 0 || tableSize > largestTableSize)
+	/*
+	 * Add the restriction qual parameter value in all baserestrictinfos.
+	 * Note that this has to be done on a copy, as the originals are needed
+	 * per target shard interval.
+	 */
+	foreach(restrictionCell, copiedRestrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *restriction = lfirst(restrictionCell);
+		Index rteIndex = restriction->index;
+		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+
+		if (!allRelationsJoinedOnPartitionKey || allReferenceTables)
```

Comment:
I'm switching to passing the restriction to the optimizer. It seems a bit weird to pass the restrictions to the optimizer, but, I agree that we should error out nicely and thus we need it.